examples/seismic/viscoelastic/viscoelastic_example.py

```diff
@@ -0,0 +1,84 @@
+import numpy as np
+from argparse import ArgumentParser
+
+from devito.logger import info
+from examples.seismic.viscoelastic import ViscoelasticWaveSolver
+from examples.seismic import demo_model, AcquisitionGeometry
+
+
+def viscoelastic_setup(shape=(50, 50), spacing=(15.0, 15.0), tn=500.,
+                       space_order=4, nbpml=10,
+                       constant=True, **kwargs):
+
+    nrec = 2*shape[0]
+    preset = 'constant-viscoelastic' if constant else 'layers-viscoelastic'
+    model = demo_model(preset, space_order=space_order, shape=shape, nbpml=nbpml,
+                       dtype=kwargs.pop('dtype', np.float32), spacing=spacing)
+
+    # Source and receiver geometries
+    src_coordinates = np.empty((1, len(spacing)))
+    src_coordinates[0, :] = np.array(model.domain_size) * .5
+    if len(shape) > 1:
+        src_coordinates[0, -1] = model.origin[-1] + 2 * spacing[-1]
+
+    rec_coordinates = np.empty((nrec, len(spacing)))
+    rec_coordinates[:, 0] = np.linspace(0., model.domain_size[0], num=nrec)
+    if len(shape) > 1:
+        rec_coordinates[:, 1] = np.array(model.domain_size)[1] * .5
+        rec_coordinates[:, -1] = model.origin[-1] + 2 * spacing[-1]
+    geometry = AcquisitionGeometry(model, rec_coordinates, src_coordinates,
+                                   t0=0.0, tn=tn, src_type='Ricker', f0=0.12)
+
+    # Create solver object to provide relevant operators
+    solver = ViscoelasticWaveSolver(model, geometry, space_order=space_order, **kwargs)
+    return solver
+
+
+def run(shape=(50, 50), spacing=(20.0, 20.0), tn=1000.0,
+        space_order=4, nbpml=40, autotune=False, constant=False, **kwargs):
+
+    solver = viscoelastic_setup(shape=shape, spacing=spacing, nbpml=nbpml, tn=tn,
+                                space_order=space_order, constant=constant, **kwargs)
+    info("Applying Forward")
+    # Define receiver geometry (spread across x, just below surface)
+    rec1, rec2, vx, vz, txx, tzz, txz, rxx, rzz, rxz, summary = \
+        solver.forward(autotune=autotune)
+
+    return rec1, rec2, vx, vz, txx, tzz, txz, rxx, rzz, rxz, summary
```

Comment:
Nitpicking, but the `r` variables are auxiliary fields to model the attenuation Q, noone really cares what they look like so don't need to ouptut it or anything. Can also ignore it in `wavesolver` and let the operator use the ones created in `operator`, should make it easier to follow, but fine with me like that if you want it.

Reply:
Sure, could tidy that up.